### PR TITLE
Fix EAGLE draft EP dispatch when input_ids is None

### LINKED ISSUE #40156: [Bug] EAGLE spec-decode: num_token_non_padded is 0 for draft batches, masking all draft topk ids to -1 and corrupting the MoE dispatch
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

Since #39574 (`0dabef3d30`, "thread `num_token_non_padded` through the MoE path"), EAGLE speculative decoding with expert parallelism can hard-crash or silently corrupt output.

The EAGLE **draft** batch reaches `ForwardBatch.init_new` with `input_ids is None`, so the new count is derived as `0` instead of the real forward width. It is `0`, not `None`, so the padded-region mask is armed with a live count of zero — and that masks **every** draft topk id to `-1`. `StandardDispatcher.dispatch` then translates that `-1` through `local_expert_mapping` *before* the drop check, and PyTorch's negative-index wrap turns it into the table's last entry, which is a real local expert on the last EP rank. The MoE then dispatches a whole batch of rows into a single expert, overrunning its `m_max` region and writing out of bounds.

Chain, with line references (all on main):

1. `ForwardBatch.init_new`: `num_tokens = len(batch.input_ids) if batch.input_ids is not None else 0` → `0` for the draft batch (the scheduler rebuilds `input_ids` after `init_new`). `global_num_token_non_padded` becomes a live `0`.
2. `TopK.forward_cuda` → `select_experts` runs the padded-region mask with live count 0 → every row of the draft `topk_ids` becomes `-1`.
3. `StandardDispatcher.dispatch`: `topk_ids_local = local_expert_mapping[topk_output.topk_ids]` runs **before** the `-1` drop check. `local_expert_mapping = torch.full((288,), -1)` with `[rank*72:(rank+1)*72] = arange(72)`, so `mapping[-1] == mapping[287] == 71` on the last EP rank (benign `-1` on ranks 0–2).
4. `masked_m[71] = 48*8 = 384 > m_max = 256`; `dst = 71*256 + [0..383] = [18176, 18559]`, past the `72*256 = 18432` region → OOB write.

Observed on the last EP rank: `masked_m[71]=384`, `src2dst=[18176,18559]`, cap `18432`.

Symptoms:
- Eager: `cudaErrorIllegalAddress`. The coredump names `fill_gateup_input_triton_kernel` (OOB write); once that write is bounded, the same corruption surfaces one stage later in `silu_mul_quant_varlen_kernel`.
- CUDA graph: no crash — silent corruption. Some in-flight requests collapse to `spec_accept_rate=0.00` and emit garbage (`!!!!!…`) up to the token cap. Requests admitted during a collapse are healthy, so the corruption is per-request persistent state.
- Disabling the overlap scheduler removes it; `CUDA_LAUNCH_BLOCKING=1` removes it.

This is a main-branch regression, not a fork artifact: the fault-path files are byte-identical to `origin/main`, and `origin/main` contains the trigger line above.

### Reproduction

EAGLE spec-decode + expert parallelism, overlap scheduler + CUDA graph, high concurrency with long generation. We hit it reliably on a DeepSeek-V2-style MoE (GLM-5.3-Flash: 288 experts, top-8, `ep_size=4` → 72 local experts/rank, EAGLE topk=1 / 3 draft steps):

```bash
python3 -m sglang.launch_server \
  --model-path <deepseek-v2-style-moe> \
  --tp 4 --ep-size 4 \
  --speculative-algorithm EAGLE --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 \
  --attention-backend dsa --moe-runner-backend deep_gemm \
  # + overlap scheduler (default) and CUDA graph
```

Then drive ~48 concurrent requests with ~8k-token generations. Eager crashes with `cudaErrorIllegalAddress`; with CUDA graph some requests collapse to `accept_rate=0.00` with `!`-flood output.

A focused repro without a full server: build a draft `ForwardBatch` with `input_ids=None` and a spec info carrying `num_tokens_per_req>0`, then check that `global_num_token_non_padded == 0` and that the resulting `topk_ids` are all `-1` (which then aliases through `local_expert_mapping`).

### Environment

From `python3 -m sglang.check_env` (image `llm/sglang:glm-dcp-0916-v3`, base `dev-cu13`):

```
Python: 3.12.3
CUDA available: True
GPU 0..7: NVIDIA H20
Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: release 13.0, V13.0.88
CUDA Driver Version: 580.178.04
PyTorch: 2.13.0+cu130
sglang: 0.0.0.dev1+g8874c51a9
sglang-kernel: 0.4.7
flashinfer_python: 0.6.18
triton: 3.7.1
transformers: 5.12.1
numpy: 2.3.5
```

Related: #39574 (merged, introduced this), #39804 (open, AMD — disables the `num_token_non_padded` threading on ROCm for the same class of failure: DSV4-Pro TP8DP8 MTP GSM8K 0.119 → 0.950; it is a platform workaround, not a fix for the EAGLE draft path).

### Suggested fixes

Upstream-grade fix (two parts), plus hardening:

1. Plumb the real token count on the EAGLE draft path — the actual trigger. In `ForwardBatch.init_new`, when `input_ids is None` for a draft/spec batch, derive the width from `spec_info.num_tokens_per_req` instead of `0`, so the padded-region mask is a no-op rather than masking every row.
2. Translate only valid ids in `StandardDispatcher.dispatch` — run the `-1` drop sentinel through unchanged instead of `local_expert_mapping[topk_ids]` (e.g. `mapping[topk_ids.clamp(min=0)]` then restore `-1` with `torch.where`), so the sentinel can never alias into a real expert regardless of the count.
3. Hardening (convert silent OOB into a loud failure): bound `fill_gateup_input_triton_kernel` and `post_reorder_deepgemm`'s gather by the `[num_local_experts, m_max, *]` region; bound `silu_mul_quant_varlen_kernel`'s row index by `num_tokens`; add an async `masked_m.max() <= T` assert on the masked-GEMM consumer side (covers the DeepEP-LL path, which has no producer-side check).

Happy to open a PR if that is useful.

### PR DESCRIPTION
**Pull request by gokay-ai**
EAGLE speculative decoding + expert parallelism has been broken on main since #39574. Draft `ForwardBatch.init_new` sees `input_ids is None` and publishes a live token count of 0, so TopK masks every draft `topk_ids` row to `-1`. `StandardDispatcher.dispatch` then indexes `local_expert_mapping[-1]`, which is a real expert on the last EP rank — illegal address in eager, silent corruption under CUDA graph.

This PR:
1. Derives the draft token width from `spec_info.num_tokens_per_req` when `input_ids` is missing.
2. Keeps the `-1` drop sentinel through `local_expert_mapping` (clamp, then restore).

Unit tests cover a draft `ForwardBatch` with `input_ids=None` and last-rank `-1` mapping. Does not change the ROCm workaround in #39804.

Fixes #40156

---
### CI States

Latest PR Test (Base): :x: [Run #35336172336](https://github.com/sgl-project/sglang/actions/runs/35336172336)
Latest PR Test (Extra): :x: [Run #35336172155](https://github.com/sgl-project/sglang/actions/runs/35336172155)
Latest PR Test (AMD ROCm 10): :x: [Run #35336172244](https://github.com/sgl-project/sglang/actions/runs/35336172244)

### CODE REVIEW

File: python/sglang/srt/layers/moe/token_dispatcher/standard.py [hunk 1]
```diff
@@ -226,15 +226,20 @@ def dispatch(
...
             elif not self.use_aiter_moe_runner:
                 if TopKOutputChecker.format_is_standard(topk_output):
-                    topk_ids_local = self.local_expert_mapping[topk_output.topk_ids]
+                    # Clamp then restore: mapping[-1] is the last table entry,
+                    # a real local expert on the last EP rank. The -1 drop
+                    # sentinel (padded-region / DP-pad mask) must not alias.
+                    topk_ids = topk_output.topk_ids
+                    topk_ids_local = self.local_expert_mapping[
+                        topk_ids.clamp(min=0)
+                    ].masked_fill(topk_ids < 0, -1)
```
Comment (JustinTong0323) [inline python/sglang/srt/layers/moe/token_dispatcher/standard.py]: Traced on current main (d94d784441): both halves of #40156 are still there, the zero draft count from ForwardBatch.init_new when input_ids is None and the unguarded local_expert_mapping[topk_ids] lookup. Clamp/lookup/restore keeps -1 as the drop sentinel without a host sync, so it is graph-safe, and the two new unit tests fail on the pre-fix sources and pass here.

What is missing before run-ci is e2e evidence: EAGLE/MTP with moe_ep_size > 1, full GSM8K with finish_reason/stop-rate plus accuracy, draft CUDA graph on and off. The draft-graph path still has no padded-row count (TODO at eagle_draft_cuda_graph_runner.py L644), so that stays out of scope here.
